src/metabase/data_apps/models.clj

```
@@ -116,24 +120,66 @@
+                                                                [:= :dad2.revision_number
+                                                                 {:select [:%max.revision_number]
+                                                                  :from   [[:data_app_definition :dad3]]
+                                                                  :where  [:= :dad3.app_id :dad.app_id]}]
+                                                                [:= :dad2.id :dad.id]]}]]}]
+
+                                  [:ranked
+                                   {:select [:dad.id :dad.app_id :dad.revision_number :dad.created_at
+                                             [[:raw "ROW_NUMBER() OVER (PARTITION BY dad.app_id ORDER BY dad.revision_number DESC)"] :app_rank]
+                                             [[:raw "ROW_NUMBER() OVER (ORDER BY dad.created_at DESC)"] :global_rank]]
+                                    :from   [[:data_app_definition :dad]]
+                                    :where  [:not [:exists {:select [1]
+                                                            :from   [[:protected_definitions :pd]]
+                                                            :where  [:= :pd.id :dad.id]}]]}]]
+
+                         :select [:id]
+                         :from   [:ranked]
+                         :where  [:or
+                                  [:> :app_rank retention-max-per-app]
+                                  [:> :global_rank retention-max-total]]}]})))
+
 (defn set-latest-definition!
   "Create a new definition for an existing app."
   [app-id definition]
-  (t2/insert-returning-instance! :model/DataAppDefinition
-                                 (merge definition
-                                        {:app_id          app-id
-                                         :revision_number (next-revision-number-hsql app-id)})))
+  (u/prog1 (t2/insert-returning-instance! :model/DataAppDefinition
```

should we do this in a transaction?

also, wdyt if we do this as part of the insert hook?

---

My knee jerk is to not to have a transaction, as the deletion operation is pretty expensive and conflict prone, so that would be heavy.

Is there a particular race condition you're thinking of?

I'm not that worried about partial failure leaving us temporarily with more history than we need, since the next insert will sort it out.

I guess we don't want to return an error if the app was created but just pruning fails. This is particularly scary if say pruning is failing due to timeout or deadlocks, and now as the user is retrying "create" we're filling the table up even more. I guess we could just catch pruning errors then, but that seems a bit low viz for self hosted instances.

As for putting it in a hook, again that feels bad to me. We don't need this for data integrity, so I'd prefer less magic.